src/terms/local.jl

```diff
@@ -71,19 +71,32 @@ function (external::ExternalFromFourier)(basis::PlaneWaveBasis{T}) where {T}
 end
 
 # Returns the form factors at unique values of |G + q| (in Cartesian coordinates).
-# Uses a hash map for O(1) lookup.
+# Additionally, returns a mapping from any G index to a unique |G + q|
 function atomic_local_form_factors(basis::PlaneWaveBasis{T}, Gqs_cart::AbstractArray) where{T}
-    form_factors = IdDict{Tuple{Int,T},T}()  # IdDict for Dual compatibility
-    for G in Gqs_cart
+    iG2norm = zeros(Int, length(Gqs_cart))
+    norms = IdDict{T, Int}()
+    for (iG, G) in enumerate(Gqs_cart)
         p = norm(G)
+        if !haskey(norms, p)
+            norms[p] = length(norms) + 1
+        end
+        iG2norm[iG] = norms[p]
+    end
+
+    form_factors = zeros(T, length(norms), length(basis.model.atom_groups))
+    for (iG, G) in enumerate(Gqs_cart)
+        p = norm(G)
+        inorm = iG2norm[iG]
         for (igroup, group) in enumerate(basis.model.atom_groups)
-            if !haskey(form_factors, (igroup, p))
+            if form_factors[inorm, igroup] ≈ zero(T)
                 element = basis.model.atoms[first(group)]
-                form_factors[(igroup, p)] = local_potential_fourier(element, p)
+                form_factors[inorm, igroup] = local_potential_fourier(element, p)
             end
         end
     end
-    form_factors
+    form_factors = to_device(basis.architecture, form_factors)
+    iG2norm = to_device(basis.architecture, iG2norm)
```

Comment:
This can lead to type instability (the type of a variable changes). The compiler may be smart enough to figure it out here, though ... but I would check.